counsel.el

```diff
@@ -49,20 +49,33 @@
 
 (define-obsolete-function-alias 'counsel-more-chars 'ivy-more-chars "0.10.0")
 
-(defun counsel-unquote-regex-parens (str)
-  "Unquote regexp parentheses in STR."
-  (if (consp str)
-      (mapconcat #'car (cl-remove-if-not #'cdr str) ".*")
-    (replace-regexp-in-string "\\\\[(){}]\\|[()]"
+(defun counsel-unquote-regex-parens (regex)
+  "Convert REGEX from Elisp format to PCRE format, on best-effort basis.
+REGEX may be of any format returned by an Ivy regex function,
+namely a string or a list.  The return value is always a string.
+
+Note that incorrect results may be returned for sufficiently
+complex regexes."
+  (if (consp regex)
+      (mapconcat
+       (lambda (pair)
+         (let ((subexp (counsel-unquote-regex-parens (car pair))))
+           (when (string-match-p "|" subexp)
+             (setq subexp (format "(%s)" subexp)))
+           subexp))
+       (cl-remove-if-not #'cdr regex)
+       ".*")
+    (replace-regexp-in-string "\\\\[(){}|]\\|[()]"
                               (lambda (s)
                                 (or (cdr (assoc s '(("\\(" . "(")
                                                     ("\\)" . ")")
                                                     ("(" . "\\(")
                                                     (")" . "\\)")
                                                     ("\\{" . "{")
-                                                    ("\\}" . "}"))))
+                                                    ("\\}" . "}")
+                                                    ("\\|" . "|"))))
```

Comment:
Nitpick: adding support for alternatives `\|` makes both the function's name and the subsequent error message "Unexpected parenthesis" confusing.

Reply:
(Having said that, the error message should only be printed in an impossible situation.)